fix: get token for active user instead of blank if possible

anuraaga opened the pull request:
Fixes #10136

I have been following this issue for a while but finally tried checking the code. It seemed pretty simple to allow the `GH_CONFIG_DIR` switching mechanism to work, and the change seems easy to reason about to me so tried sending a PR.

Note that when using `login`, we set the username-keyed token

https://github.com/cli/cli/blob/c50cdbd0c52cefc597e27c8ec4154127837e1846/internal/config/config.go#L345

so as far as I'm aware, there aren't really any gotchas, and in fact we could potentially remove the fallback to `""` - I left it for safety in case there are still some cases not covered by getting the token for the active user. If there are old versions that didn't populate the username-keyed value to keyring when logging in, then I guess they may rely on the fallback.

anuraaga committed:
fix: get token for active user instead of blank if possible

anuraaga commented on pkg/cmd/api/api_test.go:
```diff
@@ -1355,10 +1365,10 @@ func Test_apiRun_cache(t *testing.T) {
 			return &ghmock.ConfigMock{
 				AuthenticationFunc: func() gh.AuthConfig {
-					return &config.AuthConfig{}
+					return &stubAuthConfig{}
```
Since `cfg` isn't exported, I couldn't do something like `&config.AuthConfig{cfg: cfg}` which otherwise probably makes sense. Considered exporting `Cfg` but it seemed too invasive so used a stub pattern I found in another test instead.

anuraaga committed:
Cleanup

williammartin commented on internal/config/config.go:
```diff
@@ -281,6 +281,14 @@ func (c *AuthConfig) SetActiveToken(token, source string) {
 // TokenFromKeyring will retrieve the auth token for the given hostname,
 // only searching in encrypted storage.
 func (c *AuthConfig) TokenFromKeyring(hostname string) (string, error) {
+	if user, err := c.ActiveUser(hostname); err == nil && user != "" {
```
I don't have time to review this properly @andyfeller but as the person that introduced 🌚 and triaged the original issue, I'm pretty sure this is conceptually what I had in mind as a first attempt at fixing this.

This wasn't marked as `help-wanted` but I do think its _pretty_ bad, so I would welcome us prioritising this fix. I actually brought up wanting to spend some time fixing multi-account sharp edges this week with @mxie

anuraaga replied:
Thanks @williammartin - just wondering if it's possible for anyone to review this? I think there's still more that can be done to generally improve UX of multi-auth, but hopefully this fix is contained and easy enough to reason about to make some steps towards that.

williammartin replied:
Mainly just a matter of priority. We're all meeting offsite this week so there's been a lot of travelling. Would love for you to drop your thoughts on improving multi-auth UX in a new discussion. I also have lots of thoughts, and again, it's really a matter of having the evidence to prioritise it.

anuraaga replied:
Hopefully finished with offsites and such - would it be possible to get a review on this? I get priority for feature requests, but this is a bug fix so seems like it should generally be prioritized...

andyfeller replied:
Thank you, @williammartin, for picking up the slack 🙇 Having returned from offiste and trips, I'm catching up on this PR now.

anuraaga committed:
Merge branch 'trunk' into active-token-user

williammartin committed:
Use active token stubbing on auth config

williammartin commented:
## Acceptance Criteria

**Given** I have two accounts on a single github host
**And Given** I have each of these accounts as the active user in separate `hosts.yml` files
**When** I set `GH_CONFIG_DIR` to either of the files
**And When** I run `gh api /user .login`
**Then** I see the active user is correct as per the hosts.yml file

```
# Current version
➜ gh version
gh version 2.74.1 (2025-06-10)
https://github.com/cli/cli/releases/tag/v2.74.1

# auth status
➜ gh auth status --hostname github.com
github.com
  ✓ Logged in to github.com account wilmartin_microsoft (keyring)
  - Active account: true
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'gist', 'read:org', 'repo', 'workflow'

  ✓ Logged in to github.com account williammartin (keyring)
  - Active account: false
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'gist', 'read:org', 'repo', 'workflow'

# Current hosts.yml
➜ cat ~/.config/gh/hosts.yml
github.com:
    git_protocol: https
    users:
        williammartin: {}
        wilmartin_microsoft:
    user: wilmartin_microsoft

# Hit the API to see the token in use
➜ gh api /user --jq .login | cat
wilmartin_microsoft

# Adjust the hosts.yml (as if swapping GH_CONFIG_DIR)
➜  cat ~/.config/gh/hosts.yml
github.com:
    git_protocol: https
    users:
        williammartin: {}
        wilmartin_microsoft:
    user: williammartin
    
# Hit the API to see that the same (wrong token) is in use
➜ gh api /user --jq .login | cat
wilmartin_microsoft

# Build this branch
➜  cli git:(active-token-user) ✗ make
go build -trimpath -ldflags "-X github.com/cli/cli/v2/internal/build.Date=2025-06-17 -X github.com/cli/cli/v2/internal/build.Version=v2.74.0-7-g8deae3038 " -o bin/gh ./cmd/gh

# Set that it uses the correct token
➜ ./bin/gh api /user --jq .login | cat
williammartin

# And that auth status has correctly updated
➜ gh auth status --hostname github.com
github.com
  ✓ Logged in to github.com account williammartin (keyring)
  - Active account: true
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'gist', 'read:org', 'repo', 'workflow'

  ✓ Logged in to github.com account wilmartin_microsoft (keyring)
  - Active account: false
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'gist', 'read:org', 'repo', 'workflow'
 ```

williammartin reviewed (commented):
@anuraaga this is functionally correct but I _think_ that perhaps we should approach this slightly differently. I think maybe we should take the logic in `TokenFromKeyring` and push it into `ActiveToken`. I think the `ActiveToken / ActiveUser` parallel is stronger and then maybe I'll come back later and start renaming things like `TokenFromKeyring` to indicate their legacy, or fallback nature.

Is there some reason you think this wouldn't work?

anuraaga commented:
Thanks @williammartin - unless I'm missing something since this is all in `/internal/` there wouldn't even need to be renames for legacy reasons, we can just go for the end state in one shot. If the current separation is desired for readability reasons that seems fine, if pushing up is clearer that seems fine too. I can tweak this PR if there's any preference.

williammartin commented:
@anuraaga the renames would be for our maintainability. I just think I need to poke around a bit more to decide exactly what name I'd want, which is why I'm not suggesting you do it. I would like you to try and move the logic into `ActiveToken` so that it does the `ActiveUser -> TokenFromKeyringForUser -> Fallback to TokenFromKeyring`.

anuraaga committed:
Push up

anuraaga committed:
Merge branch 'trunk' of https://github.com/cli/cli into active-token-user

anuraaga commented:
Thanks @williammartin - moved the logic to `ActiveToken`, hopefully it looks ok

andyfeller committed:
Enhance Activetoken prioritize test

- ensure test user tokens are different from unkeyed token
- ensure assertion expected / actual are in correct order

williammartin commented on internal/config/auth_config_test.go:
```diff
@@ -773,7 +773,7 @@ func TestTokenPrioritizesActiveUserToken(t *testing.T) {
 	authCfg := newTestAuthConfig(t)
 	require.NoError(t, keyring.Set(keyringServiceName("github.com"), "", "test-token"))
-	require.NoError(t, keyring.Set(keyringServiceName("github.com"), "test-user1", "test-token"))
+	require.NoError(t, keyring.Set(keyringServiceName("github.com"), "test-user1", "test-token1"))
```
This isn't _really_ representative of the most common case. In nearly all cases the token in the empty username keyring entry will match one of the username keyring entries. The previous test is much more likely to be the case.

andyfeller commented:
Thank you for your patience, @anuraaga; my sincere apologies for not following up sooner! 🙇

I made a minor enhancement to one of the tests to differentiate the tokens and swapped order of test assertion.  Just seeing CI checks pass for final review.

andyfeller committed:
Restored original test setup, clarified

After discussing my previous change to the test, I'm restoring the previous keyring setup to reflect the specific situation.

I added clarifying comments to help the next reviewer.

williammartin commented:
Think this is probably ready to go @andyfeller ?

andyfeller reviewed (approved):
![You must ship it if squirrel tells you to ship it](https://github.com/user-attachments/assets/51eb9826-94d3-485c-9a75-3ba830e66674)
